scaaml/capture/aes/scope.py

```diff
@@ -0,0 +1,80 @@
+"""Context manager for the scope."""
+import chipwhisperer as cw
+
+
+class SScope:
+    """Scope context manager."""
+    def __init__(self, gain: int, samples: int, offset: int, clock: int,
+                 sample_rate: str):
+        """Create scope context.
+
+        Args:
+          gain: Gain of the scope.
+          samples: How many points to sample (length of the capture).
+          offset: Number of samples to wait before starting recording data.
+          clock: CLKGEN output frequency (in Hz).
+          sample_rate: Clock source for cw.ClockSettings.adc_src.
+        """
+        self._scope = None
+        self._gain = gain
+        self._samples = samples
+        self._offset = offset
+        self._clock = clock
+        self._sample_rate = sample_rate
+        self._basic_mode = "rising_edge"
+        self._triggers = "tio4"
+        self._tio1 = "serial_rx"
+        self._tio2 = "serial_tx"
+        self._hs2 = "clkgen"
+        self._freq_ctr_src = "clkgen"
+        self._presamples = 0
+
+    def __enter__(self):
+        """Create scope context.
+
+        Args:
+          gain: Gain of the scope.
+          samples: How many points to sample (length of the capture).
+          offset: Number of samples to wait before starting recording data.
+          clock: CLKGEN output frequency (in Hz).
+          sample_rate: Clock source for cw.ClockSettings.adc_src.
+        """
+        assert self._scope is None  # Do not allow nested with.
+        cwscope = cw.scope()
+        cwscope.gain.db = self._gain
+        max_samples = cwscope.adc.oa.hwInfo.maxSamples()
+        if self._samples > max_samples and cwscope.adc.oa.hwInfo.is_cw1200():
+            cwscope.adc.stream_mode = True
+        cwscope.adc.samples = self._samples
+        cwscope.adc.offset = self._offset
+        cwscope.adc.basic_mode = self._basic_mode
+        cwscope.clock.clkgen_freq = self._clock
+        cwscope.trigger.triggers = self._triggers
+        cwscope.io.tio1 = self._tio1
+        cwscope.io.tio2 = self._tio2
+        cwscope.io.hs2 = self._hs2
+        cwscope.clock.adc_src = self._sample_rate
+        cwscope.clock.freq_ctr_src = self._freq_ctr_src
+        cwscope.adc.presamples = self._presamples
+        self._scope = cwscope
+        return self
+
+    def __exit__(self, exc_type, exc_value, exc_tb):
+        """Safely close all resources.
+
+        Args:
+          exc_type: None if no exception, otherwise the exception type.
+          exc_value: None if no exception, otherwise the exception value.
+          exc_tb: None if no exception, otherwise the traceback.
+        """
+        # Disconnect the scope if needed
+        self._scope.dis()
```

Comment:
Is there a reason to disconnect, connect back to disable `target_pwr` (which belongs to `SControl`) to disconnect again?
I know Chipwhisperer doesn't make it easy as the scope is multifunction. `scope.io` belongs to `Scontrol`, `scope.adc`, `scope.clock`, `scope.gain` and `scope.trigger` belong to `SScope` and `scope.target` is `SCommunication`.

`scope.clock` is somehow also related to `SControl` but mainly because of the line `scope.io.hs2 = "clkgen"` which outputs the scope clock to the chip.

Reply:
Re: disconnect and connect back was copied from old code... I removed that.

Otherwise done.